### PR TITLE
Make Row and Col functions and add tests

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

**Comment by btracey** [comment]
Should I delete the Vectorer interfaces as well? They specify the previous copy/like behavior which doesn't match the current behavior. It's also not clear to me what their utility is now that rows are easily extractable. 

This just has the single fast path for Dense, which matches the previous fast path. There should be a fast path for row/column viewer, as well as the other basic types. I can either add them to this PR, or I can add a TODO.

File: mat64/dense.go [hunk 1]
```diff
@@ -231,22 +209,7 @@ func (m *Dense) SetCol(j int, src []float64) int {
...
-	}
-
-	if dst == nil {
-		dst = make([]float64, m.mat.Cols)
-	}
-	copy(dst, m.rowView(i))
-
-	return dst
-}
+// TODO(btracey): Change the comments and behavior for SetRow.
```
Comment (btracey) [inline mat64/dense.go]: Oops. I'll have to deal with this too.

**Comment by btracey** [comment]
Updates #38

**Comment by kortschak** [comment]
>Should I delete the Vectorer interfaces as well? They specify the previous copy/like behavior which doesn't match the current behavior. It's also not clear to me what their utility is now that rows are easily extractable. 

Not sure. Maybe the methods should be renamed to CopyCol, CopyRow - though I don't like these. The issue of their utility is not settled though, a matrix that is not built on a blas64 type gets a slow path, but might have very fast access to rows and columns internally. That was the intention of this type.

**Comment by kortschak** [comment]
What does this do to benchmarks?

**Comment by btracey** [comment]
We can also just change the defined behavior of Vectorer so that it returns the whole row. I agree with the plausible utility of the general type

**Comment by btracey** [comment]
(and then add a Vectorer fast path)

**Comment by kortschak** [comment]
>(and then add a Vectorer fast path)

To Row and Col? This seems like wrapping a fast path in a slow path to me.

**Comment by btracey** [comment]
The utility is that my external can take as input a Matrix, and I can extract a row using Row. If the matrix has a fast path (it is a *Dense or a Vectorer), this is automatically called during mat64.Row without the type-switch logic happening in my package code. This is the motivation for #38 in the first place.

**Comment by btracey** [comment]
Huh, running benchmarks with several -cpu values causes the tests to fail. Didn't realize the benchmark cpu flags affected the tests in any way.

**Comment by btracey** [comment]
(Looking into the cause as well, obviously)

**Comment by kortschak** [comment]
> this is automatically called during mat64.Row without the type-switch logic happening in my package code

OK, so long as it is not replacing Vectorer switching in the arithmetic code in mat64.

**Comment by btracey** [comment]
Agreed.

**Comment by kortschak** [comment]
cpu flags are test-wide, not just for benchmarks. They are important for concurrency testing.

**Comment by btracey** [comment]
Updated documentation for Vectorer and VectorSetter. This keeps the types in as they are likely useful, but documents that they should copy the full row.

**Comment by kortschak** [comment]
I'd like to know the benchmark impact before I answer this last commit. It adds a bounds check each row or column in a path that, internally, already knows the bounds are satisfied. Externally, I can understand the reasoning, but it is unfortunate for internal calls.

File: mat64/matrix.go [hunk 2]
```diff
@@ -88,32 +88,26 @@ type Mutable interface {
...
 
 // A VectorSetter can set rows and columns in the represented matrix.
 type VectorSetter interface {
-	// SetRow sets the values of the specified row to the values held in a slice of float64.
-	// It will panic if the index is out of bounds. The number of elements copied is
-	// returned and will be the minimum of the length of the slice and the number of columns
-	// in the matrix.
-	SetRow(i int, src []float64) int
+	// SetRow sets the values in the specified rows of the matrix to the values
+	// in src. len(src) must equal the numer of columns in the receiver.
```
Comment (kortschak) [inline mat64/matrix.go]: "number" (and below).

**Comment by btracey** [comment]
What do you mean by "adds a bounds check"? The Vectorer interface previously specified that the routine would panic if the index was out of bounds, so the check was there already.

**Comment by kortschak** [comment]
Ahh, sorry, I say the extra bounds check here https://github.com/gonum/matrix/pull/225/files#diff-bce3caa5fd269b104e5b15efb844e493R202 and extrapolated incorrectly.

**Comment by btracey** [comment]
BenchmarkRow using Dense of various sizes would be enough to measure? It's probably both the most common case and will have the least overhead (so that the length check plausibly matters).

**Comment by kortschak** [comment]
SGTM

**Comment by btracey** [comment]
Benchmark pushed. bound.txt is with the bounds check uncommented in the "else" branch.
````
brendan:~/Documents/mygo/src/github.com/gonum/matrix/mat64$ benchstat bound.txt nobound.txt 
name     old time/op  new time/op  delta
Row10    54.0ns ± 1%  54.7ns ± 1%  +1.23%   (p=0.000 n=9+9)
Row100   65.6ns ± 1%  66.5ns ± 2%  +1.35%   (p=0.017 n=9+9)
Row1000   210ns ± 0%   210ns ± 2%    ~     (p=0.659 n=9+10)
````
Since the benchmarks seem to suggest having the bounds check is actually faster, seems like it has negligible impact.

**Comment by kortschak** [comment]
LGTM after spelling fixed. Thanks.

**Pull request merged**